### PR TITLE
Add s3 flags to bucket cmds

### PR DESCRIPTION
**Pull request by TimSimmons**
I wanted to use these to verify buckets to see why
my compaction was potentially failing, so I added
the s3 bits. Had to refactor a little bit but it
seemed fine.

### CODE REVIEW

File: cmd/thanos/bucket.go [hunk 1]
```diff
@@ -229,3 +247,31 @@ func runBucketList(gcsBucket, format string) error {
...
 	return bkt.Iter(ctx, "", printBlock)
 }
+
+func getBucketClient(gcsBucket *string, s3config s3.Config, reg *prometheus.Registry) (objstore.Bucket, func() error, error) {
+	var (
+		bkt     objstore.Bucket
+		closeFn = func() error { return nil }
+	)
+
+	// initialize object storage
```
Comment (Bplotka) [inline cmd/thanos/bucket.go]: Please keep comments as valid sentences. (trailing dot and uppercase first letter)

File: pkg/compact/compact.go [hunk 2]
```diff
@@ -501,10 +501,6 @@ func (cg *Group) compact(ctx context.Context, dir string, comp tsdb.Compactor) (
 			}
 			uniqueSources[s] = struct{}{}
 		}
-		// Ensure all input blocks are valid.
-		if err := block.VerifyIndex(filepath.Join(pdir, "index")); err != nil {
```
Comment (Bplotka) [inline pkg/compact/compact.go]: mhm I think we need more testing for that, can you drop it from this PR to merge just bucket cmd changes (they are ok)

File: cmd/thanos/bucket.go (hunk 1 above)
Comment (TimSimmons) [reply cmd/thanos/bucket.go]: Fixed!

File: pkg/compact/compact.go (hunk 2 above)
Comment (TimSimmons) [reply pkg/compact/compact.go]: Done!

File: cmd/thanos/bucket.go [hunk 3]
```diff
@@ -56,11 +75,18 @@ func registerBucket(m map[string]setupFunc, app *kingpin.Application, name strin
...
+		bkt, closeFn, err := getBucketClient(gcsBucket, s3config, reg)
+		if err != nil {
+			return err
+		}
+
 		// Dummy actor to immediately kill the group after the run function returns.
 		g.Add(func() error { return nil }, func(error) {})
 
-		return runBucketList(*gcsBucket, *lsOutput)
+		defer closeFn()
```
Comment (Bplotka) [inline cmd/thanos/bucket.go]: this is `nil` for s3 client isn't? And will result in panic. Am I right?

Comment (TimSimmons) [reply cmd/thanos/bucket.go]: It gets defined by default as `closeFn = func() error { return nil }` on L254, should be alright.

**Comment by Bplotka** [comment]
hmm weird.. Can you rebase? Seems like you might not have fsnotify fix (https://github.com/improbable-eng/thanos/commit/2b5d828f1b7c6d7cad39bf123489f4168cc7afa1).

File: cmd/thanos/bucket.go (hunk 3 above)
Comment (Bplotka) [reply cmd/thanos/bucket.go]: Oh men.. Returning all explicitly in each case on `getBucketClient` would be more readable. (: But not a blocker.

File: cmd/thanos/bucket.go [hunk 4]
```diff
@@ -29,24 +30,42 @@ func registerBucket(m map[string]setupFunc, app *kingpin.Application, name strin
...
+	cmd.Flag("s3.access-key", "Access key for an S3-Compatible API.").
+		PlaceHolder("<key>").Envar("S3_ACCESS_KEY").StringVar(&s3config.AccessKey)
+
+	s3config.SecretKey = os.Getenv("S3_SECRET_KEY")
+
+	cmd.Flag("s3.insecure", "Whether to use an insecure connection with an S3-Compatible API.").
+		Default("false").Envar("S3_INSECURE").BoolVar(&s3config.Insecure)
+
+	cmd.Flag("s3.signature-version2", "Whether to use S3 Signature Version 2; otherwise Signature Version 4 will be used.").
+		Default("false").Envar("S3_SIGNATURE_VERSION2").BoolVar(&s3config.SignatureV2)
```
Comment (fabxc) [inline cmd/thanos/bucket.go]: Yea, we should probably add `registerS3Params` that just registers the flags and returns a config struct populated by them and the envvar.

**Comment by TimSimmons** [comment]
Updated this to make the s3 command line flags stuff a bit better (I think). The `quickstart.sh` script worked for me. Looks like the CI failure is unrelated.

**Comment by Bplotka** [comment]
Master fixed, I think you need to rebase for proper build run. (sorry)

**Comment by Bplotka** [comment]
Actually, I can see e2e were fine. Will merge - risk on me (:

**Pull request merged**